torch/utils/data/dataloader.py

```diff
@@ -246,6 +278,171 @@ def handler(signum, frame):
+    #      whenever we see an error in `next`. This isn't ideal because
+    #        a. It prevents users from using try-catch to resume data loading.
+    #        b. It doesn't prevent hanging if users have references to the
+    #           iterator.
+    #
+    #   3. All processes exit if any of them die unexpectedly (e.g., SIGKILL).
+    #
+    #      As shown above, the workers are set as daemonic children of the main
+    #      process. However, automatic cleaning-up of such child processes only
+    #      happen if the parent process exits gracefully (e.g., SIGTERM). So we
+    #      must ensure that each process will exit even the process that should
+    #      send/receive data to/from it were killed, i.e.,
+    #
+    #        a. A process won't hang when getting from a queue.
+    #
+    #           Even with carefully designed data dependencies (i.e., a `put()`
+    #           always corresponding to a `get()`), hanging on `get()` can still
+    #           happen when data in queue is corrupted (e.g., due to
+    #           `cancel_join_thread` or unexpected exit).
+    #
+    #           For child exit, we register SIGCHLD handler on main process,
+    #           and checks if any of the workers fail in the (Python) handler.
+    #           See DataLoader.cpp.
+    #
+    #           For other `.get()` calls, we guard them with timeouts, and check
+    #           the status of the sender when timeout happens. E.g., in the
+    #           workers, the `ManagerWatchdog` class is used to check main
+    #           process status.
+    #
+    #        b. A process won't hang when putting into a queue;
+    #
+    #           We use `mp.Queue` which has a separate background thread to put
+    #           objects. The background thread is usually automatically joined
+    #           when the process exits.
+    #
+    #           However, in case that the receiver has ended abruptly while
+    #           reading from the pipe, the join will hang forever. Therefore,
+    #           for both `worker_result_queue` (worker -> main process/pin_memory_thread)
+    #           and each `index_queue` (main process -> worker), we use
+    #           `q.cancel_join_thread()` in sender process before any `q.put` to
+    #           prevent this automatic join.
+    #
+    #           Note that this may leave corrupted data in the queue, but we
+    #           don't care about the data anyways once we are shutting down.
+    #
+    #
+    # Now let's get back to 1:
+    #   how we gracefully exit the workers when the last reference to the
+    #   iteartor is gone.
+    #
+    # To achieve this, we implement the following logic along with the design
+    # choices mentioned above:
+    #
+    # [pin_memory_thread] and [worker processes]
+    #   When getting from queues,
+    #     if get a `None`, exit.
+    #     if get anything else or time out, check `done_event`,
```

Comment:
On time outs we also check the watchdog, don't we?

Reply:
Yes, in case of the workers.